### ISSUE TITLE
sc.pp.filter_genes() gives error on copied data

### ISSUE THREAD

**Issue by LuckyMD** [theislab/anndata#185]
Hi,

I'm running `sc.pp.filter_genes(adata_test, min_cells=50)` and getting the error below.

This is run on a copied anndata object, and I haven't been able to reproduce is on e.g., `sc.datasets.pbmc3k`. I wonder if you have any idea what might be going wrong here though, so I can start debugging a bit smarter.

All I want to do is filter out a few genes before using `diffxpy`. Oh, and my versions are:

>scanpy==1.4.3+116.g0075c62 anndata==0.6.22.post1+29.g16615d7 umap==0.3.9 numpy==1.16.2 scipy==1.3.0 pandas==0.24.1 scikit-learn==0.20.2 statsmodels==0.10.0rc2 python-igraph==0.7.1 louvain==0.6.1

That should be fairly up to date.

```
filtered out 592 genes that are detectedin less than 50 cells
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-16-9a4043f3121d> in <module>
      4 
      5 # Filter out genes to reduce multiple testing burden
----> 6 sc.pp.filter_genes(adata_test, min_cells=50)
      7 print(f'Testing {adata_test.n_vars} genes...')
      8 print("")

~/new_scanpy/scanpy/scanpy/preprocessing/_simple.py in filter_genes(data, min_counts, min_cells, max_counts, max_cells, inplace, copy)
    218         else:
    219             adata.var['n_cells'] = number
--> 220         adata._inplace_subset_var(gene_subset)
    221         return adata if copy else None
    222 

~/new_anndata/anndata/anndata/core/anndata.py in _inplace_subset_var(self, index)
   1379         Same as ``adata = adata[:, index]``, but inplace.
   1380         """
-> 1381         adata_subset = self[:, index].copy()
   1382         self._init_as_actual(adata_subset, dtype=self._X.dtype)
   1383 

~/new_anndata/anndata/anndata/core/anndata.py in __getitem__(self, index)
   1248         """Returns a sliced view of the object."""
   1249         oidx, vidx = self._normalize_indices(index)
-> 1250         return AnnData(self, oidx=oidx, vidx=vidx, asview=True)
   1251 
   1252     def _remove_unused_categories(self, df_full, df_sub, uns):

~/new_anndata/anndata/anndata/core/anndata.py in __init__(self, X, obs, var, uns, obsm, varm, layers, raw, dtype, shape, filename, filemode, asview, oidx, vidx)
    602             if not isinstance(X, AnnData):
    603                 raise ValueError('`X` has to be an AnnData object.')
--> 604             self._init_as_view(X, oidx, vidx)
    605         else:
    606             self._init_as_actual(

~/new_anndata/anndata/anndata/core/anndata.py in _init_as_view(self, adata_ref, oidx, vidx)
    664         if adata_ref._raw is not None:
    665             # slicing along variables axis is ignored
--> 666             self._raw = adata_ref.raw[oidx]
    667         else:
    668             self._raw = None

~/new_anndata/anndata/anndata/core/anndata.py in __getitem__(self, index)
    360         new = Raw(self._adata, X=X, var=var)
    361         if self._varm is not None:
--> 362             new._varm = self._varm._view(self, (vidx,))
    363         return new
    364 

AttributeError: 'numpy.ndarray' object has no attribute '_view'
```

**Comment by LuckyMD** [comment]
I've been playing around with this a bit, and it seems it has to do with `raw` being set. I can recreate the same error with:

`adata_test = adata[[d != 'Patient1' for d in adata.obs['Donor']]].copy()`

but if I set `adata.raw = None` it all magically works. E.g.,
```
test = adata.copy()
test.raw = None
adata_test = test[[d != 'Patient1' for d in test.obs['Donor']]].copy()
```

Does slicing somehow not work when `raw` is set? Again, I haven't been able to reproduce with `sc.datasets.pbmc68k()`. No idea what makes my dataset so different...

**Comment by ivirshup** [comment]
I see what's happening with the first error, you've somehow got a numpy array for `adata.raw.varm` where it should be an `anndata.core.alignedmapping.AxisArray`. Any idea how that would have happened? Is it possible this anndata was saved via pickle?

An aside on the code, I'm not sure it actually makes sense to subset `.raw.varm` when the whole object is being subset. The `var` dimension of `.raw` should be disconnected from the `var` dimension of `adata`.

**Comment by LuckyMD** [comment]
It is indeed a `numpy.ndarray`. I saved some diffxpy objects via pickle, but never the `adata` object. That was saved via `anndata.write()` and loaded via `sc.read()`.

I actually don't care about `adata.raw` in the particular analysis I'm doing at all. So it works for me to just delete it, but this seems like a more general problem. I would agree `.raw` could stay unsliced when slicing the main object.

**Comment by LuckyMD** [comment]
@ivirshup I now have another case where I load a previously saved `adata` object and `adata.varm` is a `dict`. At the moment, every time I slice an object I have to do:
```
adata.layers['raw'] = adata.raw.X
adata.raw = None

[slice anndata object]

adata_raw = adata.copy()
adata_raw.X = adata.layers['raw']
adata.raw = adata_raw

del adata_raw
del adata.layers['raw']
```

Is there a quicker way to cast `adata.raw.varm` to `anndata.core.alignedmapping.AxisArrays`?

**Comment by ivirshup** [comment]
Yeah, this was raised in #192, and I've got a fix for it in #167. My sense is this issue would be separate from having an ndarray in `.raw.varm`. These are different files, yeah?

Right now, I think I'd just have to recommend not using `.raw.varm` if possible until I merge #167. Does scanpy have any functions which use `.raw.varm`?

If needed we could do a bug fix release to fix this, since it's an easy fix. I ended up going with adding it to my pull request because it was much easier to test with some testing infrastructure I've been adding (tests added in be1fc820a6b55e4d).

**Comment by LuckyMD** [comment]
These are different files. They are the same dataset though. One file is after I saved an analysis where I thought I had fixed the `adata.raw.varm` issue by reinstantiating `raw` as described above. It was then loaded as `dict` as you describe in #192.

I think scanpy stores PCs in `adata.varm` after `sc.pp.pca()`. Then instantiating `raw` by `adata.raw = adata` transfers that to `adata.raw.varm`. I have PCs stored there and never put anything in `varm` actively. Annoyingly you can't set `adata.raw.varm = None` either (except via a new adata object).

How far is #167 from being merged? For now I guess I can work around.

**Comment by ivirshup** [comment]
I think you could do `adata.raw._varm = None` as a work around for now.

#167 feels close, baring anything horrible coming up in review.

**Comment by LuckyMD** [comment]
Ah, cool. So I can set the attribute by referencing `raw._varm` rather than `raw.varm`?

Looking forward to the rework! Feels like views and IO playing well together will be a major upgrade.

**Comment by flying-sheep** [comment]
Fixed in https://github.com/scverse/anndata/pull/167 I assume
